src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -0,0 +1,309 @@
+    if (region->is_humongous()) {
+      if (region->get_live_data_bytes() == 0) {
+        // Humongous region is entirely garbage.  Reclaim it.
+        if (i == first_non_humongous_empty) {
+          first_non_humongous_empty++;
+        } else {
+          RegionData swap_tmp = candidates[i];
+          candidates[i] = candidates[first_non_humongous_empty];
+          candidates[first_non_humongous_empty++] = swap_tmp;
+        }
+        i++;
+      } else {
+        // Humongous region is non garbage.  Don't reclaim it.
+        if (i + 1 == first_humongous_non_empty) {
+          first_humongous_non_empty--;
+          i++;
+        } else {
+          RegionData swap_tmp = candidates[i];
+          candidates[i] = candidates[--first_humongous_non_empty];
+          candidates[first_humongous_non_empty] = swap_tmp;
+          // Do not increment i so we can revisit swapped entry on next iteration
+        }
+      }
+    } else {
+      i++;
+    }
+  }
+
+
+  // Prioritize regions to select garbage-first regions
+  QuickSort::sort<RegionData>(candidates + first_non_humongous_empty, (int)(first_humongous_non_empty - first_non_humongous_empty),
+                              compare_by_garbage, false);
+
+  // Any old-gen region that contains (ShenandoahGarbageThreshold
+  // (default value 25))% garbage or more is to be evacuated.  In the
+  // future, this threshold percentage may be specified on the command
+  // line or preferrably determined by dynamic heuristics.
+#define CollectionThresholdGarbagePercent ShenandoahGarbageThreshold
+
+  size_t region_size = ShenandoahHeapRegion::region_size_bytes();
+  for (size_t i = first_non_humongous_empty; i < first_humongous_non_empty; i++) {
+    // Do approximate percent to avoid floating point math
+    size_t percent_garbage = candidates[i]._garbage * 100 / region_size;
+
+    if (percent_garbage < CollectionThresholdGarbagePercent) {
+      _hidden_next_old_collection_candidate = 0;
+      _hidden_old_collection_candidates = i;
+      _first_coalesce_and_fill_candidate = i;
+      _old_coalesce_and_fill_candidates = cand_idx - i;
+
+      // Note that we do not coalesce and fill occupied humongous regions
+      // HR: humongous regions, RR: regular regions, CF: coalesce and fill regions
+      log_info(gc)("Old-gen mark evac (%u HR, %llu RR), %llu CF)",
+                   (unsigned int) first_non_humongous_empty,
+                   (unsigned long long) (_hidden_old_collection_candidates - first_non_humongous_empty),
+                   (unsigned long long) _old_coalesce_and_fill_candidates);
+      return;
+    }
+  }
+
+  // If we reach here, all of non-humogous old-gen regions are candidates for collection set.
+  _hidden_next_old_collection_candidate = 0;
+  _hidden_old_collection_candidates = first_humongous_non_empty;
+  _first_coalesce_and_fill_candidate = 0;
+  _old_coalesce_and_fill_candidates = 0;
+
+#undef CollectionThresholdGarbagePercent
```

Comment:
I don't think the constants need to be undef-ed when defined at top of file. Alternatively, you may consider making them true C++ constants.

Reply:
Replaced with C++ constant in next commit.